pyomo/gdp/plugins/transform_current_disjunctive_state.py

```diff
@@ -0,0 +1,211 @@
+            active=None,
+            descend_into=(Block, Disjunct),
+            sort=SortComponents.deterministic,
+        ):
+            disjunction_transform(disjunction, reverse_dict)
+
+    def _transform_disjunction(self, disjunction, reverse_dict):
+        if not disjunction.active:
+            return
+        no_val = set()
+        true_val = set()
+        false_val = set()
+        num_disjuncts = 0
+        for disjunct in disjunction.disjuncts:
+            if not disjunct.active:
+                continue
+            num_disjuncts += 1
+            ind_var = disjunct.indicator_var
+            if ind_var.value is None:
+                no_val.add(disjunct)
+            elif ind_var.value:
+                true_val.add(disjunct)
+            else:
+                false_val.add(disjunct)
+        if len(false_val) == num_disjuncts:
+            raise InfeasibleConstraintException(
+                "Logical constraint for Disjunction "
+                "'%s' is violated: All the Disjunct "
+                "indicator_vars are 'False.'" % disjunction.name
+            )
+        elif disjunction.xor:
+            if len(true_val) > 1:
+                raise InfeasibleConstraintException(
+                    "Exactly-one constraint for Disjunction "
+                    "'%s' is violated. The following Disjuncts "
+                    "are selected: %s"
+                    % (disjunction.name, ', '.join([d.name for d in true_val]))
+                )
+            elif len(true_val) == 1:
+                # We can fix everything
+                self._reclassify_disjuncts(
+                    true_val, no_val.union(false_val), reverse_dict['_disjuncts']
+                )
+                # This disjunction is fully transformed
+                reverse_dict['_disjunctions'].add(disjunction)
+                disjunction.deactivate()
+            elif len(false_val) == num_disjuncts - 1:
+                # We can fix everything. Since we didn't hit the case above, we
+                # know that the non-False value is a None.
+                self._reclassify_disjuncts(
+                    no_val, false_val, reverse_dict['_disjuncts']
+                )
+                reverse_dict['_disjunctions'].add(disjunction)
+                disjunction.deactivate()
+        # It's only an 'at-least' not an 'exactly', so if everything has a value
+        # we can transform it
+        elif len(no_val) == 0:
+            self._reclassify_disjuncts(true_val, false_val, reverse_dict['_disjuncts'])
+            reverse_dict['_disjunctions'].add(disjunction)
+            disjunction.deactivate()
+
+    def _reclassify_disjuncts(
+        self, true_disjunctions, false_disjunctions, reverse_dict
+    ):
+        for disj in true_disjunctions:
+            reverse_dict[disj] = (disj.indicator_var.fixed, disj.indicator_var.value)
+            disj.indicator_var.fix(True)
+            disj.parent_block().reclassify_component_type(disj, Block)
```

Comment:
I worry as this reclassifies *all* Disjuncts in an indexed Disjunct, whether nor not all the disjuncts state is actually determined by the `indicator_var` (e.g., only transforming part of a model)